mirror/mirror_test.go

```diff
@@ -0,0 +1,208 @@
+		// provider.
+		return ad.IsRm
+	}, testEventualTimeout, testCheckInterval, "err: %v", err)
+
+	// Read each individual mirrored ad and assert it is mirrored as expected.
+	te.requireAdChainMirroredRecursively(t, ctx, originalHeadAdCid, gotMirroredHeadAdCid)
+}
+
+func TestMirror_FormsExpectedAdChainRemap(t *testing.T) {
+	tests := []struct {
+		name          string
+		mirrorOptions []mirror.Option
+	}{
+		{
+			name: "unchanged",
+		},
+		{
+			name:          "hamt_murmur_3_3",
+			mirrorOptions: []mirror.Option{mirror.WithHamptRemapper(multihash.MURMUR3X64_64, 3, 3)},
+		},
+		{
+			name:          "hamt_id_3_1",
+			mirrorOptions: []mirror.Option{mirror.WithHamptRemapper(multihash.IDENTITY, 3, 1)},
+		},
+		{
+			name:          "entry_chunk_1",
+			mirrorOptions: []mirror.Option{mirror.WithEntryChunkRemapper(1)},
+		},
+		{
+			name:          "entry_chunk_1000",
+			mirrorOptions: []mirror.Option{mirror.WithEntryChunkRemapper(1000)},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctx := newTestContext(t)
+			rng := rand.New(rand.NewSource(testRandomSeed))
+			md := metadata.New(metadata.Bitswap{})
+
+			te := &testEnv{}
+			// Start original index provider
+			te.startSource(t, ctx, engine.WithPublisherKind(engine.DataTransferPublisher))
+
+			// Publish a bunch of ads on the original provider
+			_ = te.putAdOnSource(t, ctx, []byte("ad1"), testutil.RandomMultihashes(t, rng, 1), md)
+			_ = te.putAdOnSource(t, ctx, []byte("ad2"), testutil.RandomMultihashes(t, rng, 400), md)
+			_ = te.removeAdOnSource(t, ctx, []byte("ad1"))
+			_ = te.putAdOnSource(t, ctx, []byte("ad3"), testutil.RandomMultihashes(t, rng, 1), md)
+			_ = te.putAdOnSource(t, ctx, []byte("ad4"), testutil.RandomMultihashes(t, rng, 2), md)
+			_ = te.removeAdOnSource(t, ctx, []byte("ad2"))
+			originalHeadAdCid := te.putAdOnSource(t, ctx, []byte("ad5"), testutil.RandomMultihashes(t, rng, 7), md)
+
+			test.mirrorOptions = append(test.mirrorOptions, mirror.WithSyncInterval(time.NewTicker(time.Second)))
+			te.startMirror(t, ctx, test.mirrorOptions...)
+
+			// Await until the entire chain is mirrored; this is done by checking if the head mirrored ad
+			// is a removal.
+			var gotMirroredHeadAdCid cid.Cid
+			var err error
+			require.Eventually(t, func() bool {
+				gotMirroredHeadAdCid, err = te.mirrorSyncer.GetHead(ctx)
+				if err != nil || cid.Undef.Equals(gotMirroredHeadAdCid) {
+					return false
+				}
+				// Check the context is the latest originally published ad context as a way to
+				// assert that the entire ad chain is mirrored.
+				var ad *schema.Advertisement
+				ad, err = te.syncMirrorAd(ctx, gotMirroredHeadAdCid)
+				return err == nil && string(ad.ContextID) == "ad5"
+			}, testEventualTimeout, testCheckInterval, "err: %v", err)
+
+			// Read each individual mirrored ad and assert it is mirrored as expected.
+			te.requireAdChainMirroredRecursively(t, ctx, originalHeadAdCid, gotMirroredHeadAdCid)
+		})
+	}
+}
```

Comment:
Seems like mirroring a mirror is a possibility. Is that worth testing?

Reply:
Sure. Is there a specific scenario you have in mind that could cause issues?

Reply:
Cannot think of a specific situation. Maybe... check for correct behavior if 2nd mirror mirrors 1st mirror while 1st mirror is busy updating its content.  Make mirrors cannot mirror each other (causing a mirroring loop).

Reply:
> Make sure mirrors cannot mirror each other (causing a mirroring loop).

Not sure how we can guarantee this doesn't happen.

As for the former, I need to pick your brain to understand the scenario we are testing